services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_public.py

```diff
@@ -401,28 +403,44 @@ async def restart_containers(self, dynamic_sidecar_endpoint: AnyHttpUrl) -> None
 
 
 async def setup(app: FastAPI) -> None:
-    logger.debug("dynamic-sidecar api client setup")
-    app.state.dynamic_sidecar_api_client = DynamicSidecarClient(app)
+    with log_context(logger, logging.DEBUG, "dynamic-sidecar api client setup"):
+        app.state.dynamic_sidecar_api_clients: dict[str, DynamicSidecarClient] = {}
 
 
 async def shutdown(app: FastAPI) -> None:
-    logger.debug("dynamic-sidecar api client closing...")
-    client: Optional[DynamicSidecarClient]
-    if client := app.state.dynamic_sidecar_api_client:
-        await client._thin_client.close()  # pylint: disable=protected-access
+    with log_context(logger, logging.DEBUG, "dynamic-sidecar api client closing..."):
+        await logged_gather(
+            *(
+                x._thin_client.close()  # pylint: disable=protected-access
+                for x in app.state.dynamic_sidecar_api_clients.values()
+            )
+        )
+
+
+def get_dynamic_sidecar_client(
+    app: FastAPI, node_id: str | NodeID
+) -> DynamicSidecarClient:
+    str_node_id = f"{node_id}"
+
+    if str_node_id not in app.state.dynamic_sidecar_api_clients:
+        app.state.dynamic_sidecar_api_clients[str_node_id] = DynamicSidecarClient(app)
+
+    return app.state.dynamic_sidecar_api_clients[str_node_id]
+
 
+def remove_dynamic_sidecar_client(app: FastAPI, node_id: str | NodeID) -> None:
```

Comment:
use only NodeID in the interface? would make the code more straightforward no?

Reply:
I actually wanted for the user to have the option to pass both `str` and `NodeID`.